napari/utils/perf/_trace_file.py

```diff
@@ -0,0 +1,92 @@
+"""PerfTraceFile class to write JSON files in the chrome://tracing format.
+"""
+import json
+import os
+import threading
+
+from ._compat import perf_counter_ns
+from ._event import PerfEvent
+
+
+class PerfTraceFile:
+    """Writes a chrome://tracing formatted JSON file.
+
+    Chrome has a nice built-in performance tool called chrome://tracing. Chrome
+    can record traces of web applications. But the format is well-documented and
+    anyone can create the files and use the nice GUI. And other programs accept
+    the format including:
+    1) https://www.speedscope.app/ which does flamegraphs (Chrome doesn't).
+    2) Qt Creator's performance tools.
+
+    Attributes
+    ----------
+    zero_ns : int
+        perf_counter_ns() time when we started the trace.
+    pid : int
+        Process ID.
+    tid : int
+        Thread ID.
+    outf : file handle
+        JSON file we are writing to.
+
+    Notes
+    -----
+    There are two chrome://tracing formats:
+    1) JSON Array Format
+    2) JSON Object Format
+
+    We are using the JSON Array Format right now, the file can be cut off at
+    anytime. The other format allows for more options if we need them but must
+    be closed properly.
+
+    See the "trace_event format" Google Doc for details:
+    https://chromium.googlesource.com/catapult/+/HEAD/tracing/README.md
+    """
+
+    def __init__(self, path: str):
+        """Open the tracing file on disk.
+
+        Parameters
+        ----------
+        path : str
+            Write the trace file to this path.
+        """
```

Comment:
side note here:  I can't find the conversations where this has been discussed at the moment, but for better or worse, we follow the [numpy convention](https://numpydoc.readthedocs.io/en/latest/format.html#class-docstring) in the napari codebase, which is to put the full docstring for the init method in `Parameters` section in the class docstring.

Reply:
Didn't realize that was the convention. Fixed.